persist/fs/commitlog/reader.go

```diff
@@ -110,52 +156,121 @@ func (r *reader) Read() (
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
+
+func (r *reader) readLoop() {
+	index := 0
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			return
+		default:
+			data, err := r.readChunk()
+			// TODO: Special case if err == io.EOF here?
+
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
+		}
 	}
+}
 
-	if len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan *readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := &readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			outBuf <- readResponse
+			continue
+		}
+		decoder.Reset(arg.bytes)
+		entry, err := decoder.DecodeLogEntry()
 		if err != nil {
-			resultErr = err
-			return
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
 		}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		r.metadataLookup[entry.Index] = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       uint32(decoded.Shard),
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
+
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadataLookupLock.Lock()
+			r.metadataLookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       uint32(decoded.Shard),
+			}
+			r.metadataLookupLock.Unlock()
+			namespace.DecRef()
+			id.DecRef()
 		}
-	}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+		numMetadataMiss := 0
+		var metadata Series
+		var ok bool
+		for true {
+			r.metadataLookupLock.Lock()
+			metadata, ok = r.metadataLookup[entry.Index]
+			r.metadataLookupLock.Unlock()
+
+			if ok {
+				break
+			}
+
+			// Race condition where the metadata is currently being processed by
```

Comment:
I really don't like sleeping in here, but I couldn't think of a better way to synchronize the map.

The performance loss due to the sleep is negligible because this doesn't happen that often (less than 20 times per file)